### PR TITLE
St/mempool account size

### PR DESCRIPTION
**Pull request by saratnt**

### CODE REVIEW

File: sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala [hunk 1]
```diff
@@ -101,14 +100,45 @@ class MempoolMap(
...
+    if (accountSize + txSizeInSlot(ethTransaction) > MaxSlotsPerAccount) {
+      log.trace(s"Adding transaction $ethTransaction exceeds maximum allowed size per account")
+      throw AccountMemPoolOutOfBoundException(ethTransaction.id)
+    }
+    all.put(ethTransaction.id, ethTransaction)
+    txByNonceMap.put(ethTransaction.getNonce, ethTransaction.id)
+  }
+
+  def getAccountSize(account: SidechainTypes#SCP): Long = {
+    val execSize: Long = executableTxs.get(account).map(executableTxsPerAccount => executableTxsPerAccount.values.foldLeft(0L) { (sum, txId) => sum + txSizeInSlot(all(txId)) }).getOrElse(0L)
```
Comment (i-Alex) [inline sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala:115]: 'txSizeInSlot' is based on `tx.size()` method.
every time when executed tx is serialized from to bytes and then returns the `size`.
I think we need to cache `size` value for EthereumTransaction.

File: sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala [hunk 2]
```diff
@@ -267,13 +300,20 @@ class MempoolMap(
     var destMap = newExecTxs
     var haveBecomeNonExecutable = false
     var maxNonceGapExceeded = false
+    var currAccountSlots = 0L
```
Comment (i-Alex) [inline sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala:303]: do we need to use `Long` type here ?
same question about `txSizeInSlot` method.
`Int` must be more than enough.

**Review by i-Alex — CHANGES_REQUESTED** [review]
Good!

File: qa/sc_evm_mempool_invalid_txs.py [hunk 3]
```diff
@@ -248,7 +252,34 @@ def run_test(self):
 
         assert_true(exception_occurs, "Adding a transaction with nonce too low should fail")
         response = allTransactions(sc_node_1, False)
-        assert_equal(0, len(response["transactionIds"]), "Transaction with nonce too low  added to node 1 mempool")
+        assert_equal(0, len(response["transactionIds"]), "Transaction with nonce too low added to node 1 mempool")
+
+        # Test that a transaction exceeding the account size is rejected by the mem pool
+        #Create as many txs of 1 slot as to almost fill the mempool
+        for i in range(9, 9 + (self.max_account_slots - 1)):
```
[the remark this answers is not in the document]
Comment (saratnt) [reply qa/sc_evm_mempool_invalid_txs.py:259]: Done

File: sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala (hunk 2 above)
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala:303]: This is already solved in PR: https://github.com/HorizenOfficial/Sidechains-SDK/pull/728

File: sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala (hunk 1 above)
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala:115]: Done

**Review by alsala — CHANGES_REQUESTED** [review]

File: qa/sc_evm_mempool_invalid_txs.py [hunk 4]
```diff
@@ -211,11 +213,12 @@ def run_test(self):
                      "Transaction that creates a smart contract with empty data added to node 1 mempool")
 
         # Test that a transaction with nonce gap too big is rejected by the mem pool
+        too_big_nonce = self.max_nonce_gap + 10
         exception_occurs = False
         try:
             createEIP1559Transaction(sc_node_1, fromAddress=evm_address_sc1, toAddress=None,
-                                          nonce=100, gasLimit=10, maxPriorityFeePerGas=900000000,
-                                            maxFeePerGas=900000000, value=1, data='012344')
+                                     nonce=too_big_nonce, gasLimit=10, maxPriorityFeePerGas=900000000,
```
[the remark this answers is not in the document]
Comment (saratnt) [reply qa/sc_evm_mempool_invalid_txs.py:220]: Done in previous PR

File: sdk/src/main/java/com/horizen/account/transaction/EthereumTransaction.java [hunk 5]
```diff
@@ -283,7 +285,10 @@ public void semanticValidity() throws TransactionSemanticValidityException {
 
     @Override
     public long size() {
```
[the remark this answers is not in the document]
Comment (saratnt) [reply sdk/src/main/java/com/horizen/account/transaction/EthereumTransaction.java:287]: I don't think it is really necessary but for uniformity with other similar methods in EthereumTransaction I added `synchronized`

File: sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala [hunk 6]
```diff
@@ -101,14 +100,45 @@ class MempoolMap(
     this
   }
 
-  def replaceIfCanPayHigherFee(existingTxId: ModifierId, newTx: SidechainTypes#SCAT, mapOfTxsByNonce: TxIdByNonceMap) = {
+  private[mempool] def addNewTransaction(txByNonceMap: TxIdByNonceMap, ethTransaction: SidechainTypes#SCAT) = {
+    val account = ethTransaction.getFrom
+    val accountSize = getAccountSize(account)
```
[the remark this answers is not in the document]
Comment (saratnt) [reply sdk/src/main/scala/com/horizen/account/mempool/MempoolMap.scala:105]: Done

**Review by alsala — APPROVED** [review]

**Review by i-Alex — APPROVED** [review]

**Pull request merged**